sklearn/tree/_tree.pyx

```diff
@@ -1056,6 +1056,28 @@ cdef class Tree:
 
         return out
 
+    cpdef compute_node_depths(self):
+        """Compute the depth of each node in a tree.
+
+        Returns
+        -------
+        depths : ndarray of shape (self.node_count,), dtype=np.int64
+            The depth of each node in the tree.
+
+        """
+        cdef cnp.int64_t[::1] depths = np.empty(self.node_count, dtype=np.int64)
+        cdef cnp.npy_intp node_id
+        cdef cnp.int64_t depth
+
+        # init root node
+        depths[0] = 1
+        for node_id in range(self.node_count):
+            if self.children_left[node_id] != _TREE_LEAF:
+                depth = depths[node_id] + 1
+                depths[self.children_left[node_id]] = depth
+                depths[self.children_right[node_id]] = depth
+
+        return depths.base
```

Comment:
When I did the prototyping, it was super important to get a memoryview of the children array. Otherwise, there is too many Python interaction. I assume that there is an issue with some indirections. On the top of the head doing the following should provide a x100 speed-up.

```suggestion
        cdef:
            cnp.int64_t[::1] depths = np.empty(self.node_count, dtype=np.int64)
            cnp.npy_intp[:] children_left = self.children_left
            cnp.npy_intp[:] children_right = self.children_right
            cnp.npy_intp node_id
            cnp.npy_intp node_count = self.node_count
            cnp.int64_t depth

        # init root node
        depths[0] = 1
        for node_id in range(node_count):
            if children_left[node_id] != _TREE_LEAF:
                depth = depths[node_id] + 1
                depths[children_left[node_id]] = depth
                depths[children_right[node_id]] = depth

        return depths.base
```